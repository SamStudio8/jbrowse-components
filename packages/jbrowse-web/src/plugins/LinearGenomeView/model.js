import shortid from 'shortid'
import { types, getParent, getSnapshot } from 'mobx-state-tree'
import { ConfigurationSchema } from '../../util/configuration'

const IdType = types.optional(types.identifier, shortid.generate)

export const Block = types
  .model({
    refName: types.string,
    start: types.number,
    end: types.number,
  })
  .views(self => ({
    get widthPx() {
      return Math.abs(self.end - self.start) / getParent(self, 2).bpPerPx
    },
  }))

const minTrackHeight = 20
export const Track = types.model('Track', {
  id: IdType,
  name: types.string,
  type: types.string,
  height: types.optional(
    types.refinement('trackHeight', types.number, n => n >= minTrackHeight),
    minTrackHeight,
  ),
  subtracks: types.maybe(types.array(types.late(() => Track))),
  configuration: ConfigurationSchema('Track', {
    backgroundColor: {
      description: `the track's background color`,
      type: 'color',
      defaultValue: '#eee',
    },
  }),
})

const ViewStateBase = types.model({
  // views have an auto-generated ID by default
  id: IdType,
})

const LinearGenomeViewState = types
  .compose(
    'LinearViewModel',
    ViewStateBase,
    types.model({
      type: types.literal('linear'),
      offsetPx: 0,
      bpPerPx: 1,
      blocks: types.array(Block),
      tracks: types.array(Track),
      controlsWidth: 100,
      width: 800,
      configuration: ConfigurationSchema('LinearView', {
        bar: { type: 'integer', model: types.integer, defaultValue: 0 },
      }),
    }),
  )
  .views(self => ({
    get totalBlocksWidthPx() {
      return self.blocks.reduce((a, b) => a + b.widthPx, 0)
    },
  }))
  .actions(self => ({
    addTrack(id, name, type) {
      self.tracks.push(Track.create({ id, name, type }))
    },

    pushBlock(refName, start, end) {
      self.blocks.push(Block.create({ refName, start, end }))
    },

    horizontalScroll(distance) {
      self.offsetPx = Math.min(
        Math.max(self.offsetPx + distance, 0),
        self.width - self.totalBlocksWidthPx - self.controlsWidth,
      )
    },
  }))

export default LinearGenomeViewState
