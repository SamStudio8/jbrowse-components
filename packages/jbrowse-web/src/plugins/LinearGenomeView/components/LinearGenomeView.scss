.LinearGenomeView {
    background: #eee;
    border: 1px solid hsl(0, 0%, 0%);
    box-sizing: content-box;

    .view-controls {
      height: 100%;
      border-bottom: 1px solid #9e9e9e;
      box-sizing: border-box;
    }

    .block {
      &.left-side {
        border-left: 2px solid #333
      }
      &.right-side {
        border-right: 2px solid #333
      }
    }

    .controls {
      border-right: 1px solid gray;
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    }
    .track-controls {
      white-space: normal;
    }

    .TrackRenderingContainer {
      overflow: hidden;
      background: #333;
      white-space: nowrap;


      .TrackRenderingNotImplemented {
        color: white;
        padding: 0 1em;
        text-align: center;
      }
    }

    .drag-handle {
      background: #ccc;
      cursor: ns-resize;
      width: 100%;
      box-sizing: border-box;
      border-top: 1px solid #fafafa;
    }
}
