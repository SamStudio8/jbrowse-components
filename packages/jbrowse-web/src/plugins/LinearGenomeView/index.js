import modelFactory from './models'
import ReactComponent from './components/LinearGenomeView'

import Plugin from '../../Plugin'
import ViewType from '../../pluggableElementTypes/ViewType'

import configSchema from './models/configSchema'
import BasicTrackFactory from './BasicTrack'
import DynamicTrackFactory from './DynamicTrack'
import TrackType from '../../pluggableElementTypes/TrackType'

export default class LinearGenomeViewPlugin extends Plugin {
  install(pluginManager) {
    pluginManager.addTrackType(() => {
      const { stateModel, configSchema: trackConfig } = BasicTrackFactory(
        pluginManager,
      )

      return new TrackType({
        name: 'BasicTrack',
        configSchema: trackConfig,
        stateModel,
      })
    })

    pluginManager.addTrackType(() => {
      const { stateModel, configSchema: trackConfig } = DynamicTrackFactory(
        pluginManager,
      )

      return new TrackType({
        name: 'DynamicTrack',
        configSchema: trackConfig,
        stateModel,
      })
    })

    pluginManager.addViewType(() => {
      this.installed = true

      return new ViewType({
        name: 'LinearGenomeView',
        stateModel: modelFactory(pluginManager),
        configSchema,
        ReactComponent,
      })
    })
  }
}
