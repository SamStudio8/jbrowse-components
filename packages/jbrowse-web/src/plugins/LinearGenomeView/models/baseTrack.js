import React from 'react'
import { types, getRoot, getParent } from 'mobx-state-tree'
import {
  ConfigurationSchema,
  ConfigurationReference,
} from '../../../configuration'
import { ElementId } from '../../../mst-types'

import TrackControls from '../components/TrackControls'

export const BaseTrackConfig = ConfigurationSchema('BaseTrack', {
  viewType: 'LinearGenomeView',
  name: {
    description: 'descriptive name of the track',
    type: 'string',
    defaultValue: 'Track',
  },
  description: {
    description: 'a description of the track',
    type: 'string',
    defaultValue: '',
  },
  category: {
    description: 'the category and sub-categories of a track',
    type: 'stringArray',
    defaultValue: [],
  },
})

// these MST models only exist for tracks that are *shown*.
// they should contain only UI state for the track, and have
// a reference to a track configuration (stored under root.configuration.tracks).

// note that multiple displayed tracks could use the same configuration.

// note that multiple displayed tracks could use the same configuration.
const minTrackHeight = 20
const BaseTrack = types
  .model('BaseTrack', {
    id: ElementId,
    type: types.string,
    height: types.optional(
      types.refinement('trackHeight', types.number, n => n >= minTrackHeight),
      minTrackHeight,
    ),
    subtracks: types.literal(undefined),
    configuration: ConfigurationReference(BaseTrackConfig),
  })
  .views(self => ({
    get ControlsComponent() {
      return TrackControls
    },

    get RenderingComponent() {
      return (
        self.reactComponent ||
        (() => (
          <div className="TrackRenderingNotImplemented">
            Rendering not implemented for {self.type} tracks
          </div>
        ))
      )
    },

    get renderProps() {
      // view -> [tracks] -> [blocks]
      const view = getParent(self, 2)
      return {
        bpPerPx: view.bpPerPx,
        horizontallyFlipped: view.horizontallyFlipped,
      }
    },
  }))
  .actions(self => ({
    setHeight(trackHeight) {
      if (trackHeight >= minTrackHeight) self.height = trackHeight
    },

    activateConfigurationUI() {
      getRoot(self).editConfiguration(self.configuration)
    },
  }))

export default BaseTrack
