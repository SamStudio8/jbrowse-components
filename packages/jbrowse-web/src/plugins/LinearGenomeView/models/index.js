import { transaction } from 'mobx'
import {
  getEnv,
  getRoot,
  isStateTreeNode,
  types,
  getType,
  getParent,
} from 'mobx-state-tree'
import { getConf, readConfObject } from '../../../configuration'
import { ElementId, Region } from '../../../mst-types'
import { clamp } from '../../../util'
import PluginManager from '../../../PluginManager'
import TrackType from '../../../pluggableElementTypes/TrackType'

import LinearGenomeViewConfigSchema from './configSchema'

import BaseTrack from './baseTrack'
import calculateBlocks from './calculateBlocks'

// these MST models only exist for tracks that are *shown*.
// they should contain only UI state for the track, and have
// a reference to a track configuration (stored under root.configuration.tracks).

const ViewStateBase = types.model({
  // views have an auto-generated ID by default
  id: ElementId,
})

const minBpPerPx = 0.03

export default function LinearGenomeViewStateFactory(pluginManager) {
  return types
    .compose(
      'LinearGenomeView',
      ViewStateBase,
      types.model({
        id: ElementId,
        type: types.literal('LinearGenomeView'),
        offsetPx: 0,
        bpPerPx: 1,
        flipped: false,
        // we use an array for the tracks because the tracks are displayed in a specific
        // order that we need to keep.
        tracks: types.array(
          pluginManager.pluggableMstType('track', 'stateModel'),
        ),
        controlsWidth: 120,
        displayedRegions: types.array(Region),
        configuration: LinearGenomeViewConfigSchema,
      }),
    )
    .views(self => ({
      get width() {
        return getEnv(self).testEnv ? 800 : getRoot(self).viewsWidth
      },
      get totalBlocksWidthPx() {
        return self.blocks.reduce((a, b) => a + b.widthPx, 0)
      },
      get maxBpPerPx() {
        const displayWidth = self.width - self.controlsWidth
        let totalbp = 0
        self.displayedRegions.forEach(region => {
          totalbp += region.end - region.start
        })
        return totalbp / displayWidth
      },
      get minBpPerPx() {
        return minBpPerPx
      },

      /**
       * calculate the blocks we should be showing
       */
      get blocks() {
        return calculateBlocks(self, self.horizontallyFlipped)
      },

      get horizontallyFlipped() {
        return getConf(self, 'reversed')
      },
    }))
    .actions(self => ({
      showTrack(configuration, initialSnapshot = {}) {
        const { type } = configuration
        const name = readConfObject(configuration, 'name')
        const trackType = pluginManager.getTrackType(type)
        if (!trackType) throw new Error(`unknown track type ${type}`)
        const track = trackType.stateModel.create(
          Object.assign({}, initialSnapshot, { name, type, configuration }),
        )
        self.tracks.push(track)
      },

      hideTrack(configuration) {
        // if we have any tracks with that configuration, turn them off
        const shownTracks = self.tracks.filter(
          t => t.configuration === configuration,
        )
        transaction(() => shownTracks.forEach(t => self.tracks.remove(t)))
        return shownTracks.length
      },

      closeView() {
        getParent(self, 2).removeView(self)
      },

      toggleTrack(configuration) {
        // if we have any tracks with that configuration, turn them off
        const hiddenCount = self.hideTrack(configuration)
        // if none had that configuration, turn one on
        if (!hiddenCount) self.showTrack(configuration)
      },

      displayRegions(regions) {
        self.displayedRegions = regions.map(r =>
          isStateTreeNode(r) ? r : Region.create(r),
        )
      },

      activateTrackSelector() {
        if (getType(self.configuration).name === 'AnonymousModel')
          throw new Error('this view should have a real configuration')
        const trackSelectorType = getConf(self, 'trackSelectorType')
        if (trackSelectorType === 'hierarchical') {
          const rootModel = getRoot(self)
          if (!rootModel.drawerWidgets.get('hierarchicalTrackSelector'))
            rootModel.addDrawerWidget(
              'HierarchicalTrackSelectorDrawerWidget',
              'hierarchicalTrackSelector',
              { view: self },
            )
          const selector = rootModel.drawerWidgets.get(
            'hierarchicalTrackSelector',
          )
          selector.setView(self)
          rootModel.setTask('track_select', self)
          rootModel.showDrawerWidget(selector)
        } else {
          throw new Error(`invalid track selector type ${trackSelectorType}`)
        }
      },

      zoomTo(newBpPerPx) {
        if (newBpPerPx === self.bpPerPx) return
        let bpPerPx = newBpPerPx
        if (bpPerPx < self.minBpPerPx) bpPerPx = self.minBpPerPx
        else if (bpPerPx > self.maxBpPerPx) bpPerPx = self.maxBpPerPx
        self.bpPerPx = bpPerPx
      },

      resizeTrack(trackId, distance) {
        const track = self.tracks.find(t => t.id === trackId)
        if (track) track.setHeight(track.height + distance)
      },

      horizontalScroll(distance) {
        const leftPadding = 10
        const rightPadding = 10
        const displayRegionsTotalPx = self.displayedRegions.reduce(
          (a, b) => a + (b.end - b.start) / self.bpPerPx,
          0,
        )
        const maxOffset = displayRegionsTotalPx - leftPadding
        const displayWidth = self.width - self.controlsWidth
        const minOffset = -displayWidth + rightPadding
        self.offsetPx = clamp(self.offsetPx + distance, minOffset, maxOffset)
      },

      activateConfigurationUI() {
        getRoot(self).editConfiguration(self.configuration)
      },
    }))
}

// a stub linear genome view state model that only accepts base track types.
// used in unit tests.
const stubManager = new PluginManager()
stubManager.addTrackType(
  () =>
    new TrackType({
      name: 'Base',
      stateModel: BaseTrack,
      RenderingComponent: true,
    }),
)
stubManager.configure()
export const TestStub = LinearGenomeViewStateFactory(stubManager)
