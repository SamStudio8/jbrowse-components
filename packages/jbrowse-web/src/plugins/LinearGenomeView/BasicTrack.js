import { types, getParent } from 'mobx-state-tree'
import blockBasedTrack from './models/blockBasedTrack'
import {
  ConfigurationReference,
  ConfigurationSchema,
} from '../../configuration'
import BlockBasedTrackComponent from './components/BlockBasedTrack'
import { BaseTrackConfig } from './models/baseTrack'

export default pluginManager => {
  const configSchema = ConfigurationSchema(
    'BasicTrack',
    {
      adapter: pluginManager.pluggableConfigSchemaType('adapter'),
      rendering: pluginManager.pluggableConfigSchemaType('renderer'),
    },
    { baseConfiguration: BaseTrackConfig, explicitlyTyped: true },
  )

  const stateModel = types.compose(
    'BasicTrack',
    blockBasedTrack,
    types
      .model({
        type: types.literal('BasicTrack'),
        configuration: ConfigurationReference(configSchema),
        height: 100,
      })
      .views(self => ({
        get renderProps() {
          const view = getParent(self, 2)
          return {
            bpPerPx: view.bpPerPx,
            horizontallyFlipped: view.horizontallyFlipped,
            trackModel: self,
            config: self.configuration.rendering,
          }
        },

        get rendererTypeName() {
          return self.configuration.rendering.type
        },
      }))
      .volatile(() => ({
        reactComponent: BlockBasedTrackComponent,
      })),
  )

  return { stateModel, configSchema }
}
