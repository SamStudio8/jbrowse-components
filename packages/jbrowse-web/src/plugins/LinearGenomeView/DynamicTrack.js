import { types } from 'mobx-state-tree'

import BasicTrack from './BasicTrack'
import {
  ConfigurationSchema,
  ConfigurationReference,
} from '@gmod/jbrowse-core/configuration'

export default pluginManager => {
  const {
    stateModel: basicTrackStateModel,
    configSchema: basicTrackConfigSchema,
  } = BasicTrack(pluginManager)

  const configSchema = ConfigurationSchema(
    'DynamicTrack',
    {},
    { baseConfiguration: basicTrackConfigSchema, explicitlyTyped: true },
  )

  // a DynamicTrack is just a BasicTrack but with blockType hardcoded to 'dynamicBlocks'
  const stateModel = types.compose(
    'DynamicTrack',
    basicTrackStateModel,
    types
      .model({
        type: types.literal('DynamicTrack'),
        configuration: ConfigurationReference(configSchema),
      })
      .views((/* self */) => ({
        get blockType() {
          return 'dynamicBlocks'
        },
      })),
  )

  return { stateModel, configSchema }
}
