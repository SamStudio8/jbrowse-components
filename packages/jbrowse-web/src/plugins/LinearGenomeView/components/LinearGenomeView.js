import React, { Component } from 'react'
import { values } from 'mobx'
import { inject, observer, PropTypes } from 'mobx-react'

import ScaleBar from './ScaleBar'
import TrackRenderingContainer from './TrackRenderingContainer'
import TrackBlocks from './TrackBlocks'
import TrackResizeHandle from './TrackResizeHandle'

import './LinearGenomeView.scss'

const dragHandleHeight = 3

@inject('rootModel')
@observer
class LinearGenomeView extends Component {
  static propTypes = {
    model: PropTypes.observableObject.isRequired,
  }

  static wrappedComponent = {
    propTypes: {
      rootModel: PropTypes.observableObject.isRequired,
    },
  }

  render() {
    const scaleBarHeight = 22
    const { model, rootModel } = this.props
    const {
      id,
      blocks,
      tracks,
      bpPerPx,
      width,
      controlsWidth,
      offsetPx,
    } = model
    // NOTE: offsetPx is the total offset in px of the viewing window into the
    // whole set of concatenated regions. this number is often quite large.
    // visibleBlocksOffsetPx is the offset of the viewing window into the set of blocks
    // that are *currently* being displayed
    const visibleBlocksOffsetPx = blocks[0] ? offsetPx - blocks[0].offsetPx : 0
    const height =
      scaleBarHeight +
      values(tracks)
        .filter(t => t.visible)
        .reduce((a, b) => a + b.height + dragHandleHeight, 0)
    const style = {
      width: `${width}px`,
      height: `${height}px`,
      gridTemplateRows: `[scale-bar] auto ${values(tracks)
        .filter(t => t.visible)
        .map(
          t => `[${t.id}] ${t.height}px [resize-${t.id}] ${dragHandleHeight}px`,
        )
        .join(' ')}`,
      gridTemplateColumns: `[controls] ${controlsWidth}px [blocks] auto`,
    }
    return (
      <div className="LinearGenomeView" key={`view-${id}`} style={style}>
        <div
          className="controls view-controls"
          style={{ gridRow: 'scale-bar' }}
        >
          <button
            type="button"
            onClick={() => {
              if (!rootModel.drawerWidgets.get(id))
                rootModel.addDrawerWidget(
                  'HierarchicalTrackSelectorDrawerWidget',
                  id,
                )
              rootModel.showDrawerWidget(id)
            }}
          >
            select tracks
          </button>
        </div>
        <ScaleBar
          style={{ gridColumn: 'blocks', gridRow: 'scale-bar' }}
          height={scaleBarHeight}
          bpPerPx={bpPerPx}
          blocks={blocks}
          offsetPx={visibleBlocksOffsetPx}
          width={width - controlsWidth}
        />
        {values(tracks)
          .filter(t => t.visible)
          .map(track => [
            <div
              className="controls track-controls"
              key={`controls:${track.id}`}
              style={{ gridRow: track.id, gridColumn: 'controls' }}
            >
              {track.name || track.id}
            </div>,
            <TrackRenderingContainer
              key={`track-rendering:${track.id}`}
              trackId={track.id}
              width={width - controlsWidth}
              onHorizontalScroll={model.horizontalScroll}
            >
              <track.RenderingComponent
                model={track}
                blockDefinitions={blocks}
                offsetPx={visibleBlocksOffsetPx}
                bpPerPx={bpPerPx}
                blockState={{}}
              />
            </TrackRenderingContainer>,
            <TrackResizeHandle
              key={`handle:${track.id}`}
              trackId={track.id}
              onVerticalDrag={model.resizeTrack}
            />,
          ])}
      </div>
    )
  }
}

export default LinearGenomeView
