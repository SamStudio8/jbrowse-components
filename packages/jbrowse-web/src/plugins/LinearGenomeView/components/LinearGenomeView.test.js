import { Provider } from 'mobx-react'
import React from 'react'
import renderer from 'react-test-renderer'
import { TestStub as Model } from '../models'
import LinearGenomeView from './LinearGenomeView'

describe('LinearGenomeView genome view component', () => {
  it('renders with an empty model', () => {
    const model = Model.create(
      {
        type: 'LinearGenomeView',
        offsetPx: 0,
        bpPerPx: 1,
        tracks: [],
        controlsWidth: 100,
        configuration: {},
      },
      {
        testEnv: true,
      },
    )
    const component = renderer.create(
      <Provider rootModel={{ activeDrawerWidgets: new Map() }}>
        <LinearGenomeView model={model} />
      </Provider>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders one track, no blocks', () => {
    const model = Model.create(
      {
        type: 'LinearGenomeView',
        offsetPx: 0,
        bpPerPx: 1,
        tracks: [
          {
            id: 'foo',
            name: 'Foo Track',
            type: 'tester',
            height: 20,
            configuration: { name: 'foo' },
          },
        ],
        controlsWidth: 100,
        configuration: {},
      },
      {
        testEnv: true,
      },
    )
    const component = renderer.create(
      <Provider rootModel={{ activeDrawerWidgets: new Map() }}>
        <LinearGenomeView model={model} />
      </Provider>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders two tracks, two regions', () => {
    const model = Model.create(
      {
        type: 'LinearGenomeView',
        offsetPx: 0,
        bpPerPx: 1,
        displayedRegions: [
          { assemblyName: 'volvox', refName: 'ctgA', start: 0, end: 100 },
          { assemblyName: 'volvox', refName: 'ctgB', start: 1000, end: 200 },
        ],
        tracks: [
          {
            id: 'foo',
            type: 'tester',
            height: 20,
            configuration: { name: 'foo' },
          },
          {
            id: 'bar',
            name: 'Bar Track',
            type: 'tester',
            height: 20,
            configuration: { name: 'bar' },
          },
        ],
        controlsWidth: 100,
      },
      {
        testEnv: true,
      },
    )
    const component = renderer.create(
      <Provider rootModel={{ activeDrawerWidgets: new Map() }}>
        <LinearGenomeView model={model} />
      </Provider>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
