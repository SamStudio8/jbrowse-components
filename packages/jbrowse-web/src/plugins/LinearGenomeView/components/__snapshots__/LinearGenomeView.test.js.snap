// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinearGenomeView genome view component renders one track, no blocks 1`] = `
<div
  className="LinearGenomeView"
  style={
    Object {
      "gridTemplateColumns": "[controls] 100px [blocks] auto",
      "gridTemplateRows": "[scale-bar] auto [foo] 20px [resize-foo] 3px",
      "height": "45px",
      "width": "800px",
    }
  }
>
  <div
    className="controls view-controls"
    style={
      Object {
        "gridRow": "scale-bar",
      }
    }
  >
    <button
      onClick={[Function]}
      type="button"
    >
      select tracks
    </button>
  </div>
  <div
    className="ScaleBar"
    style={
      Object {
        "gridColumn": "blocks",
        "gridRow": "scale-bar",
        "height": "22px",
        "width": "700px",
      }
    }
  />
  <div
    className="controls track-controls"
    style={
      Object {
        "gridColumn": "controls",
        "gridRow": "foo",
      }
    }
  >
    Foo Track
  </div>
  <div
    className="TrackRenderingContainer"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onWheel={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "blocks",
        "gridRow": "foo",
        "width": "700px",
      }
    }
  >
    <div
      className="TrackRenderingNotImplemented"
    >
      Rendering not implemented for 
      tester
       tracks
    </div>
  </div>
  <div
    className="drag-handle drag-handle-horizontal"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "span 2",
        "gridRow": "resize-foo",
      }
    }
  />
</div>
`;

exports[`LinearGenomeView genome view component renders two tracks, two blocks 1`] = `
<div
  className="LinearGenomeView"
  style={
    Object {
      "gridTemplateColumns": "[controls] 100px [blocks] auto",
      "gridTemplateRows": "[scale-bar] auto [foo] 20px [resize-foo] 3px [bar] 20px [resize-bar] 3px",
      "height": "68px",
      "width": "800px",
    }
  }
>
  <div
    className="controls view-controls"
    style={
      Object {
        "gridRow": "scale-bar",
      }
    }
  >
    <button
      onClick={[Function]}
      type="button"
    >
      select tracks
    </button>
  </div>
  <div
    className="ScaleBar"
    style={
      Object {
        "gridColumn": "blocks",
        "gridRow": "scale-bar",
        "height": "22px",
        "width": "700px",
      }
    }
  />
  <div
    className="controls track-controls"
    style={
      Object {
        "gridColumn": "controls",
        "gridRow": "foo",
      }
    }
  >
    Foo Track
  </div>
  <div
    className="TrackRenderingContainer"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onWheel={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "blocks",
        "gridRow": "foo",
        "width": "700px",
      }
    }
  >
    <div
      className="TrackRenderingNotImplemented"
    >
      Rendering not implemented for 
      tester
       tracks
    </div>
  </div>
  <div
    className="drag-handle drag-handle-horizontal"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "span 2",
        "gridRow": "resize-foo",
      }
    }
  />
  <div
    className="controls track-controls"
    style={
      Object {
        "gridColumn": "controls",
        "gridRow": "bar",
      }
    }
  >
    Bar Track
  </div>
  <div
    className="TrackRenderingContainer"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onWheel={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "blocks",
        "gridRow": "bar",
        "width": "700px",
      }
    }
  >
    <div
      className="TrackRenderingNotImplemented"
    >
      Rendering not implemented for 
      tester
       tracks
    </div>
  </div>
  <div
    className="drag-handle drag-handle-horizontal"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "span 2",
        "gridRow": "resize-bar",
      }
    }
  />
</div>
`;

exports[`LinearGenomeView genome view component renders with an empty model 1`] = `
<div
  className="LinearGenomeView"
  style={
    Object {
      "gridTemplateColumns": "[controls] 100px [blocks] auto",
      "gridTemplateRows": "[scale-bar] auto [foo] 20px [resize-foo] 3px [bar] 20px [resize-bar] 3px [baz] 20px [resize-baz] 3px",
      "height": "91px",
      "width": "800px",
    }
  }
>
  <div
    className="controls view-controls"
    style={
      Object {
        "gridRow": "scale-bar",
      }
    }
  >
    <button
      onClick={[Function]}
      type="button"
    >
      select tracks
    </button>
  </div>
  <div
    className="ScaleBar"
    style={
      Object {
        "gridColumn": "blocks",
        "gridRow": "scale-bar",
        "height": "22px",
        "width": "700px",
      }
    }
  />
  <div
    className="controls track-controls"
    style={
      Object {
        "gridColumn": "controls",
        "gridRow": "foo",
      }
    }
  >
    Foo Track
  </div>
  <div
    className="TrackRenderingContainer"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onWheel={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "blocks",
        "gridRow": "foo",
        "width": "700px",
      }
    }
  >
    <div
      className="TrackRenderingNotImplemented"
    >
      Rendering not implemented for 
      tester
       tracks
    </div>
  </div>
  <div
    className="drag-handle drag-handle-horizontal"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "span 2",
        "gridRow": "resize-foo",
      }
    }
  />
  <div
    className="controls track-controls"
    style={
      Object {
        "gridColumn": "controls",
        "gridRow": "bar",
      }
    }
  >
    Bar Track
  </div>
  <div
    className="TrackRenderingContainer"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onWheel={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "blocks",
        "gridRow": "bar",
        "width": "700px",
      }
    }
  >
    <div
      className="TrackRenderingNotImplemented"
    >
      Rendering not implemented for 
      tester
       tracks
    </div>
  </div>
  <div
    className="drag-handle drag-handle-horizontal"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "span 2",
        "gridRow": "resize-bar",
      }
    }
  />
  <div
    className="controls track-controls"
    style={
      Object {
        "gridColumn": "controls",
        "gridRow": "baz",
      }
    }
  >
    Baz Track
  </div>
  <div
    className="TrackRenderingContainer"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    onWheel={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "blocks",
        "gridRow": "baz",
        "width": "700px",
      }
    }
  >
    <div
      className="TrackRenderingNotImplemented"
    >
      Rendering not implemented for 
      tester
       tracks
    </div>
  </div>
  <div
    className="drag-handle drag-handle-horizontal"
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    role="presentation"
    style={
      Object {
        "gridColumn": "span 2",
        "gridRow": "resize-baz",
      }
    }
  />
</div>
`;
